---
title: 'The HDF5 ecosystem in Bioconductor'
tags:
  - R
  - Bioconductor
  - HDF5
  - out-of-memory
authors:
  - name: Mike L. Smith
    orcid: 0000-0002-7800-3848
    equal-contrib: true
    corresponding: true
    affiliation: 1
  - name: Hervé Pagès
    equal-contrib: true
    affiliation: 2
affiliations:
 - name: European Molecular Biology Laboratory (EMBL), Genome Biology Unit, Heidelberg, Germany
   index: 1
 - name: Fred Hutchinson Cancer Research Center, Seattle, Washington, United States of America
   index: 2
date: 20 March 2023
bibliography: paper.bib

# Optional fields if submitting to a AAS journal too, see this blog post:
# https://blog.joss.theoj.org/2018/12/a-new-collaboration-with-aas-publishing
#aas-doi: 10.3847/xxxxx <- update this with the DOI from AAS once you know it.
#aas-journal: Astrophysical Journal <- The name of the AAS journal.
---

# Summary

Bioconductor is an open-source software project and community that provides
tools for the analysis and comprehension of biological data. One of the key
missions of Bioconductor is to provide software infrastructure for common tasks
and datatypes, allowing researchers to focus on developing interoperable tools
for higher-level data analysis without reinventing the wheel.  As the volume of
data produced by experiments in molecular biology continues to grow,
outstripping advances in compute power, it has become imperative to provide
technical solutions that allow researchers to easily interact with datasets that
exceed the memory capacity of standard consumer hardware.  One such approach is
provided by the HDF5 (Hierarchical Data Format 5) file format, which allows
efficient access to subsets of compressed data on disk. Here we present the
current state-of-the-art for working with HDF5 files in Bioconductor.

# Statement of need

The HDF5 file format has emerged as a popular approach for storing the vast
amounts of data generated by modern molecular biology research. In particular,
for single-cell sequencing experiments, 10X Genomics will provide raw data as
HDF5 files.  Additionally, there have emerged 3 major, competing,
representations of single-cell experiments (Seurat [@Satija2015],
AnnData [@Wolf2018SCANPYLS], SingleCellExperiment [@Amezquita2019OrchestratingSA]) each of which can be backed by an HDF5 dataset.  Providing a
performant, well documented, and user-friendly R interface to access these files
is a cornerstone to Bioconductor’s strategy for handling large-scale biological
datasets.

## Rhdf5lib

The Rhdf5lib package provides a distribution of the HDF5 library as an R
package. It serves as the bedrock of Bioconductors' HDF5 support, with the
other packages described in this article linking against it.  Providing the HDF5
library in this format simplifies the otherwise potentially complicated
installation process across all supported operating systems, ensuring that
the toolchain compatible with a users R version is also used to compile HDF5.
This approach also provides a consistent version of the HDF5 API for Bioconductor developers to target.

## rhdf5filters

One of the key features of the HDF5 file format is the ability to pass chunked
data through user-defined filters during one the way to or from disk, typically
to modify or compress a dataset.  HDF5 includes several filter algorithms as
standard, and the version of the HDF5 library found in Rhdf5lib is compiled with
support for the zlib and aec [@aec2020] compression filters which rely on
third-party compression libraries.  However in some usecases, depending on the
nature of the data in question, an alternative compression algorithm may be the
most appropriate to use. To enable this HDF5 is able to use dynamically loaded
filters, which are compiled independently from the HDF5 library, and are
available to an application at run time.  The rhdf5filters package makes use of
this functionality, distributing several external HDF5 filters that enable the
use of bzip2, lzf, or the Blosc meta-compressor when performing I/O operations
on HDF5 files in Bioconductor.


## rhdf5

rhdf5 provides the ability to interact with HDF5 files.  

Firstly, it provides
a large majority of the functionality found in the HDF5 C-API, maintaining
as much of the functionality found there as possible.

In addition rhdf5 also provides simplified 'wrapper' functions for common operations (e.g.
writing a new dataset), with many of the more 

To illustrate the two approaches the code below demonstrates how an rhdf5 user could
create a new HDF5 file and populate it was a dataset called "A" that contains 
the two integers 1 & 2.

```r
# C API-like interface
## Create and HDF5 file, dataspace, and dataset
fid <- H5Fcreate( name = "/my/special/file.h5" )
sid <- H5Screate_simple( c(2,1) )
did <- H5Dcreate( fid, "A", "H5T_STD_I32LE", sid )

## write a vector of length 2 to the dataset
H5Dwrite(did, 1L:2L, h5spaceMem = sid, h5spaceFile = sid)

## close all open HDF5 object handles
H5Dclose( did )
H5Sclose( sid )
H5Fclose( fid )

# Simplified wrapper functions
## Create an HDF5 file
h5createFile( file = "/my/special/file.h5" )
## Write a vector of length 2 to a dataset call "A"
h5write( file = "/my/special/file.h5",
         obj = 1L:2L,
         name = "A" )
```

## HDF5array

The previously mentioned packages provide general purpose tools for working with
HDF5 files in R, regardless of the research domain or datatype.  However, for users
of Bioconductor (and indeed R in general) it is common to work with single
HDF5 datasets representing vectors, matrices, or higher-dimensional arrays.
The HDF5array package defines R classes and methods that provide convenient
array-like containers for representing and manipulating HDF5 datasets in a memory-efficient manner, using syntax that is familiar to any R programmer.


# Availability

All four packages can be readily installed from Bioconductor. Developmental
versions can be found on GitHub.   Code contributions, bug reports, fixes and
feature requests are most welcome by opening issues and pull requests at the
appropriate GitHub repositories.

<!--
# Citations

Citations to entries in paper.bib should be in
[rMarkdown](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html)
format.

If you want to cite a software repository URL (e.g. something on GitHub without a preferred
citation) then you can do it with the example BibTeX entry below for @fidgit.

For a quick reference, the following citation commands can be used:
- `@author:2001`  ->  "Author et al. (2001)"
- `[@author:2001]` -> "(Author et al., 2001)"
- `[@author1:2001; @author2:2001]` -> "(Author1 et al., 2001; Author2 et al., 2002)"
-->
# Figures
<!--
# Figures can be included like this:
# ![Caption for example figure.\label{fig:example}](figure.png)
# and referenced from text using \autoref{fig:example}.
# 
# Figure sizes can be customized by adding an optional second parameter:
# ![Caption for example figure.](figure.png){ width=20% }
-->
# Acknowledgements

We would like to acknowledge the significant contribution of Bernd Fischer, the
original author of rhdf5.  In addition, we also thank Gregoire Pau, Martin
Morgan, and Daniel van Twisk for their contributions to the various pacakges.

# References