---
title: 'The HDF5 ecosystem in Bioconductor'
tags:
  - R
  - Bioconductor
  - HDF5
  - out-of-memory
authors:
  - name: Mike L. Smith
    orcid: 0000-0002-7800-3848
    equal-contrib: true
    corresponding: true
    affiliation: 1
  - name: Hervé Pagès
    equal-contrib: true
    affiliation: 2
affiliations:
 - name: European Molecular Biology Laboratory (EMBL), Genome Biology Unit, Heidelberg, Germany
   index: 1
 - name: Fred Hutchinson Cancer Research Center, Seattle, Washington, United States of America
   index: 2
date: 20 March 2023
bibliography: paper.bib

# Optional fields if submitting to a AAS journal too, see this blog post:
# https://blog.joss.theoj.org/2018/12/a-new-collaboration-with-aas-publishing
#aas-doi: 10.3847/xxxxx <- update this with the DOI from AAS once you know it.
#aas-journal: Astrophysical Journal <- The name of the AAS journal.
---

# Summary



# Statement of need

The HDF5 file format has emerged as a popular approach for storing the vast
amounts of data generated by modern molecular biology research. In particular,
for single-cell sequencing experiments, 10X Genomics will provide raw data as
HDF5 files.  Additionally, there have emerged 3 major, competing,
representations of single-cell experiments (Seurat [@Satija2015],
AnnData [@Wolf2018SCANPYLS], SingleCellExperiment [@Amezquita2019OrchestratingSA]) each of which can be backed by an HDF5 dataset.  Providing a
performant, well documented, and user-friendly R interface to access these files
is a cornerstone to Bioconductor’s strategy for handling large-scale biological
datasets.

## Rhdf5lib

The Rhdf5lib package provides a distribution of the HDF5 library as an R
package. It serves as the bedrock of Bioconductors' HDF5 support, with the
other packages described in this article linking against it.  Providing the HDF5
library in this format simplifies the otherwise potentially complicated
installation process across all supported operating systems, ensuring that
the toolchain compatible with a users R version is also used to compile HDF5.
This approach also provides a consistent version of the HDF5 API for Bioconductor developers to target.

## rhdf5filters

One of the key features of the HDF5 file format is the ability to pass chunked
data through user-defined filters during one the way to or from disk, typically
to modify or compress a dataset.  HDF5 includes several filter algorithms as
standard, and the version of the HDF5 library found in Rhdf5lib is compiled with
support for the zlib and aec [@aec2020] compression filters which rely on
third-party compression libraries.  However in some usecases, depending on the
nature of the data in question, an alternative compression algorithm may be the
most appropriate to use. To enable this HDF5 is able to use dynamically loaded
filters, which are compiled independently from the HDF5 library, and are
available to an application at run time.  The rhdf5filters package makes use of
this functionality, distributing several external HDF5 filters that enable the
use of bzip2, lzf, or the Blosc meta-compressor when performing I/O operations
on HDF5 files in Bioconductor.


## rhdf5

## HDF5array

# Availability

All four packages can be readily installed from Bioconductor. Developmental
versions can be found on GitHub. 
<!--# Package dependencies and system requirements
are documented at # https://js2264.github.io/VplotR/. VplotR has been tested
using R (version 3.5 
# and later) on macOS (versions 10.11 and later), Ubuntu
18.04.2 and Windows 
# machines. To ensure stability, VplotR includes unit tests
for most functions and 
# supports continuous integration using the Travis CI
platform. -->
Code contributions, bug reports, fixes and feature requests are most
welcome by opening issues and pull requests at the appropriate GitHub
repositories.

<!--
# Citations

Citations to entries in paper.bib should be in
[rMarkdown](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html)
format.

If you want to cite a software repository URL (e.g. something on GitHub without a preferred
citation) then you can do it with the example BibTeX entry below for @fidgit.

For a quick reference, the following citation commands can be used:
- `@author:2001`  ->  "Author et al. (2001)"
- `[@author:2001]` -> "(Author et al., 2001)"
- `[@author1:2001; @author2:2001]` -> "(Author1 et al., 2001; Author2 et al., 2002)"
-->
# Figures
<!--
# Figures can be included like this:
# ![Caption for example figure.\label{fig:example}](figure.png)
# and referenced from text using \autoref{fig:example}.
# 
# Figure sizes can be customized by adding an optional second parameter:
# ![Caption for example figure.](figure.png){ width=20% }
-->
# Acknowledgements

We acknowledge significant contributions from Bernd Fischer, the original author
of rhdf5.

# References